xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -145,15 +171,191 @@ type watchUpdate struct {
-	cancelWatch    func()
-	edsLB          balancer.Balancer
+	ccw            *ccWrapper            // ClientConn interface passed to child LB.
+	bOpts          balancer.BuildOptions // BuildOptions passed to child LB.
+	updateCh       *buffer.Unbounded     // Channel for gRPC and xdsClient updates.
+	xdsClient      xdsClientInterface    // xDS client to watch Cluster resource.
+	cancelWatch    func()                // Cluster watch cancel func.
+	edsLB          balancer.Balancer     // EDS child policy.
 	clusterToWatch string
 	logger         *grpclog.PrefixLogger
 	closed         *grpcsync.Event
+
+	// The certificate providers are cached here to that they can be closed when
+	// a new provider is to be created.
+	cachedRoot     certprovider.Provider
+	cachedIdentity certprovider.Provider
+	xdsHI          *xds.HandshakeInfo
+	xdsCredsInUse  bool
+}
+
+// handleClientConnUpdate handles a ClientConnUpdate received from gRPC. Good
+// updates lead to registration of a CDS watch. Updates with error lead to
+// cancellation of existing watch and propagation of the same error to the
+// edsBalancer.
+func (b *cdsBalancer) handleClientConnUpdate(update *ccUpdate) {
+	// We first handle errors, if any, and then proceed with handling the
+	// update, only if the status quo has changed.
+	if err := update.err; err != nil {
+		b.handleErrorFromUpdate(err, true)
+	}
+	if b.xdsClient == update.client && b.clusterToWatch == update.clusterName {
+		return
+	}
+	if update.client != nil {
+		// Since the cdsBalancer doesn't own the xdsClient object, we don't have
+		// to bother about closing the old client here, but we still need to
+		// cancel the watch on the old client.
+		b.cancelWatch()
+		b.xdsClient = update.client
+	}
+	if update.clusterName != "" {
+		cancelWatch := b.xdsClient.WatchCluster(update.clusterName, b.handleClusterUpdate)
+		b.logger.Infof("Watch started on resource name %v with xds-client %p", update.clusterName, b.xdsClient)
+		b.cancelWatch = func() {
+			cancelWatch()
+			b.logger.Infof("Watch cancelled on resource name %v with xds-client %p", update.clusterName, b.xdsClient)
+		}
+		b.clusterToWatch = update.clusterName
+	}
+}
+
+// handleSecurityConfig processes the security configuration received from the
+// management server, creates appropriate certificate provider plugins, and
+// updates the HandhakeInfo which is added as an address attribute in
+// NewSubConn() calls.
+func (b *cdsBalancer) handleSecurityConfig(config *xdsclient.SecurityConfig) error {
+	// If xdsCredentials are not in use, i.e, the user did not want to get
+	// security configuration from an xDS server, we should not be acting on the
+	// received security config here. Doing so poses a security threat.
+	if !b.xdsCredsInUse {
+		return nil
+	}
+	// Security config being nil is a valid case where the management server has
+	// not sent any security configuration. The xdsCredentials implementation
+	// handles this by delegating to its fallback credentials.
```

Comment:
Where does that happen?  It looks like we're eating any config updates that say to go back to using the fallback creds.

Reply:
When the user specifies xds creds, but the `Cluster` resource does not contain any security configuration, we need to use the fallback config, and the way to do that is by basically doing nothing here.

Not sure if I understood your question though.

Reply:
What if the first update says to use zatar and a subsequent update wants to undo that and revert to the fallback creds?  Is that not allowed?

Reply:
Yes, good catch. I seem to have not handled it. This requires a small change in the whole flow:

- `ccWrapper` would always add a `xds.HandshakeInfo` to the addresses in NewSubConn. This `xds.HandshakeInfo` would reflect the current state. The fields would be `nil` if no config has been received so far. The fields would also be `nil` in the case where a good config was received at one point, but was followed by an empty config. The fields would be `non-nil` when the last received config contained security information.
- The `xdsCreds` code would revert to fallback when all fields in the `HandshakeInfo` are `nil`, and not when `HandshakeInfo` itself is nil (as this will not be possible if we do the above change).

With this change, we will be able to cleanly handle both transitions (from fallback to zatar and vice-versa). What do you think?

Reply:
I have implemented the changes mentioned above.